spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/integration/AbstractTraceChannelInterceptor.java

```diff
@@ -29,33 +30,42 @@ protected AbstractTraceChannelInterceptor(TraceManager traceManager) {
 	 * trace id passed initially.
 	 */
 	Span buildSpan(Message<?> message) {
-		String spanId = getHeader(message, Trace.SPAN_ID_NAME);
-		String traceId = getHeader(message, Trace.TRACE_ID_NAME);
-		if (StringUtils.hasText(traceId)) {
-			MilliSpan.MilliSpanBuilder span = MilliSpan.builder().traceId(traceId).spanId(spanId);
-			String parentId = getHeader(message, Trace.PARENT_ID_NAME);
-			if (message.getHeaders().containsKey(Trace.NOT_SAMPLED_NAME)) {
-				span.exportable(false);
-			}
-			String processId = getHeader(message, Trace.PROCESS_ID_NAME);
-			String spanName = getHeader(message, Trace.SPAN_NAME_NAME);
-			if (spanName != null) {
-				span.name(spanName);
-			}
-			if (processId != null) {
-				span.processId(processId);
-			}
-			if (parentId != null) {
-				span.parent(parentId);
-			}
-			span.remote(true);
-			return span.build();
+		if (!hasHeader(message, Trace.TRACE_ID_NAME) || !hasHeader(message, Trace.SPAN_ID_NAME)) {
+			return null;
+		}
+		long spanId = hasHeader(message, Trace.SPAN_ID_NAME) ?
+				getHeader(message, Trace.SPAN_ID_NAME, Long.class) : new Random().nextLong();
+		long traceId = getHeader(message, Trace.TRACE_ID_NAME, Long.class);
+		MilliSpan.MilliSpanBuilder span = MilliSpan.builder().traceId(traceId).spanId(spanId);
+		Long parentId = getHeader(message, Trace.PARENT_ID_NAME, Long.class);
+		if (message.getHeaders().containsKey(Trace.NOT_SAMPLED_NAME)) {
+			span.exportable(false);
+		}
+		String processId = getHeader(message, Trace.PROCESS_ID_NAME);
+		String spanName = getHeader(message, Trace.SPAN_NAME_NAME);
+		if (spanName != null) {
+			span.name(spanName);
+		}
+		if (processId != null) {
+			span.processId(processId);
+		}
+		if (parentId != null) {
+			span.parent(parentId);
 		}
-		return null;
+		span.remote(true);
+		return span.build();
 	}
 
 	String getHeader(Message<?> message, String name) {
-		return (String) message.getHeaders().get(name);
+		return getHeader(message, name, String.class);
+	}
+
+	<T> T getHeader(Message<?> message, String name, Class<T> type) {
```

Comment:
suggestion, if you make this getHeaderAsLong.. you are down to 2 args, and slightly less complex looking call sites